### PR TITLE
feat: Column sources for ui.table formatting

### LINKED ISSUE #984: ui.table format values from a column
We should support reading the value for a format from a column. We could do this in a few ways

1. Check if all values are column names. If they are, add the column to `always_fetch_columns`
2. Add a `*_col` field for the values
3. Have a dataclass that has a key to indicate a value is a column. E.g. `color=ui.TableColumn("X")` and then the `TableColumn` dataclass has metadata indicating it's a column.

### PR DESCRIPTION
**Pull request by mattrunyon**
Fixes #984

Allows all string format values to instead specify a column as the source. Ensures the column is always fetched and throws if the column is not a string type (can cause some really bad performance issues if value formatting is invalid). Also modified to just resolve all theme colors since this change could lead to unresolved theme colors in a column source.

The example in the docs or this example shows the feature working. Change the String cast on line 6 to non-string to test the error is thrown and displayed.

This example uses an input table joined to the table so you can adjust the formatting colors via input table (something Raffi asked about specifically)

```py
from deephaven import input_table
from deephaven import ui
import deephaven.plot.express as dx

_stocks = dx.data.stocks()
_source = _stocks.select_distinct("Sym").update("Color=(String)null")

color_source = input_table(init_table=_source, key_cols="Sym")

t = ui.table(
    _stocks.natural_join(color_source, "Sym", "SymColor=Color"),
    hidden_columns=["SymColor"],
    format_=[
        ui.TableFormat(cols="Sym", background_color="SymColor")
    ],
)
```

### CODE REVIEW

File: plugins/ui/src/js/src/elements/UITable/UITable.tsx [hunk 1]
```diff
@@ -39,9 +46,11 @@ const log = Log.module('@deephaven/js-plugin-ui/UITable');
  * @returns A stable array if none of the elements have changed
  */
 function useStableArray<T>(array: T[]): T[] {
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  const stableArray = useMemo(() => array, [...array]);
-  return stableArray;
+  const stableArray = useRef<T[]>(array);
+  if (!array.every((v, i) => v === stableArray.current[i])) {
```
Comment (bmingles) [inline plugins/ui/src/js/src/elements/UITable/UITable.tsx:50]: Curious what this improves over the previous version?

File: plugins/ui/src/js/src/elements/UITable/UITableModel.ts [hunk 2]
```diff
@@ -417,20 +417,39 @@ class UITableModel extends IrisGridModel {
...
+      let resolvedFormatValue = formatValue;
+      const columnSourceIndex =
+        typeof formatValue === 'string'
+          ? this.getColumnIndexByName(formatValue)
+          : null;
+      if (columnSourceIndex != null) {
+        const columnSource = this.columns[columnSourceIndex];
+        if (!TableUtils.isStringType(columnSource.type)) {
+          throw new Error(
+            `Column ${columnSource.name} used as a formatting value for ${formatKey} is type ${columnSource.type}. Must be a String`
```
Comment (bmingles) [inline plugins/ui/src/js/src/elements/UITable/UITableModel.ts:430]:
```suggestion
            `Column ${columnSource.name} which provides TableFormat values for ${formatKey} is of type ${columnSource.type}. Columns that provide TableFormat values must be of type string.`
```

**Review by bmingles — CHANGES_REQUESTED** [review]
1 suggestion on error message. Otherwise, code looks good. Verified the String + non-string cases.

File: plugins/ui/docs/components/table.md [hunk 3]
```diff
@@ -66,6 +66,24 @@ t = ui.table(
 )
 ```
 
+### Formatting value from a column
```
[the remark this answers is not in the document]
Comment (mattrunyon) [reply plugins/ui/docs/components/table.md]: Changed to "Formatting values from a column source"

**Review by margaretkennedy — DISMISSED** [review]

**Review by bmingles — CHANGES_REQUESTED** [review]
1 more suggestion to optimize the array memoization

**Review by bmingles — APPROVED** [review]
LGTM

**Review by margaretkennedy — APPROVED** [review]

**Pull request merged**